veg2hab/functionele_samenhang.py

```diff
@@ -0,0 +1,295 @@
+
+    Afhankelijk van het percentage van een complexdeel wordt er gebufferd, en daarna adhv overlap in de GeoDataFrame clusters bepaald;
+    Als A overlapt met B en B met C, dan is [A, B, C] een cluster
+
+    Output is een lijst met lijsten van (ElmID, complexdeelindex), waarbij iedere sublijst een cluster is,
+    en er geen gemene elementen zijn tussen sublijsten.
+    """
+    assert (
+        "identifier" in gdf.columns
+    ), "identifier moet een kolom zijn in de gegeven GeoDataFrame"
+    assert (
+        "percentage" in gdf.columns
+    ), "percentage moet een kolom zijn in de gegeven GeoDataFrame"
+    assert (
+        "habtype" in gdf.columns
+    ), "habtype moet een kolom zijn in de gegeven GeoDataFrame"
+    assert (
+        "geometry" in gdf.columns
+    ), "geometry moet een kolom zijn in de gegeven GeoDataFrame"
+
+    distance_tuples = (
+        Interface.get_instance().get_config().functionele_samenhang_buffer_distances
+    )
+    intersection_pairs = set()
+    for distance_tuple in distance_tuples:
+        min_perc, buffer_distance = distance_tuple
+        current_subset = gdf[gdf.percentage >= min_perc].copy()
+        current_subset["geometry"] = current_subset.buffer(buffer_distance)
+
+        # Bepalen overlap-paren
+        overlaps = gpd.sjoin(
+            current_subset, current_subset, how="inner", predicate="intersects"
+        )
+        overlaps = list(
+            zip(
+                overlaps["identifier_left"].tolist(),
+                overlaps["identifier_right"].tolist(),
+            )
+        )
+        intersection_pairs.update(overlaps)
+
+    clusters = UnionFind.cluster_pairs(intersection_pairs)
+
+    # Alle elementen die niet aan de laagste percentage-eis voldoen moeten als zelfstandige nog steeds wel
+    # meegenomen worden in de rest van het minimum-oppervlak-proces, dus die voegen we als individuen toe
+    lowest_perc = min(distance_tuples, key=lambda x: x[0])[0]
+    unclusterables = gdf[gdf.percentage < lowest_perc]
+    clusters.extend(
+        [[identifier] for identifier in unclusterables["identifier"].tolist()]
+    )
+
+    return clusters
+
+
+def _extract_elmid_perc_habtype(gdf: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
+    """
+    Trekt uit de vegkartering gdf de benodigde informatie voor het bepalen van clusters
+    Ieder complexdeel/HabitatKeuze krijgt een eigen rij in de output, met de identifier (ElmID, complex-deel-index)
+    Als er meerdere complexdelen met hetzelfde habtype in een vlak zitten, worden deze samengevoegd
+    """
+
+    def apply_func(row: pd.Series) -> pd.DataFrame:
+        # identifier (Elmid, [cmplxdeel_n]) | percentage | habitattype | geometry
+        identifier = []
+        percentage = []
+        habtype = []
+        geometry = []
+        for idx, keuze in enumerate(row["HabitatKeuze"]):
+            # Dit stelt ons in staat weer terug te gaan naar de originele habitatkeuze
+            identifier.append((row["ElmID"], (idx,)))
```

Comment:
Ik denk dat het geen kwaad kan, wat vaker NameTuples te gebruiken, dat maakt dit soort zaken wat leesbaarder.

Reply:
Ik zat al te denken "moet ik hier een class van maken" maar dat voelde wat overkill, maar namedtuple is een goeie tussenoplossing (ook al zittie dus nog niet zo goed in mn systeem)